framework/src/postprocessors/GeneralSensorPostprocessor.C

```diff
@@ -0,0 +1,315 @@
+    Real _for_int_at_delay = linearInterpolationInVectors(__time_values, __for_int, delay_value);
+    // push the delay and interpolated signal value into the new vector
+    __for_int.push_back(_for_int_at_delay);
+    __time_values.push_back(delay_value);
+    _for_exp.push_back(_for_exp_at_delay);
+    // sort the vector in aascending order
+    std::sort(__for_int.begin(), __for_int.end());
+    std::sort(__time_values.begin(), __time_values.end());
+    std::sort(_for_exp.begin(), _for_exp.end());
+    // New vector for integration with all elements before delay removed from __for_int and _for_exp
+    removeValuesBeforeTime(__time_values, __for_int, _for_exp, delay_value); 
+
+    integrand = elementwiseMultiply(__for_int, _for_exp);
+    //LinearInterpolation object with two vectors
+    LinearInterpolation integral(__time_values, integrand);
+    //integrate the vectors
+    integration_value = integral.integrate(); 
+  }
+  
+  else if (_t > delay_value && delay_value < _dt)
+  {
+    // linear interpolation of for_int at delay time
+    Real _for_int_at_delay = linearInterpolation(0, __time_values[0], for_int_initial, __for_int[0], delay_value);
+    // push the delay and interpolated signal value into the new vector
+    __for_int.push_back(_for_int_at_delay);
+    __time_values.push_back(delay_value);
+    _for_exp.push_back(_for_exp_at_delay);
+    // sort the vector in aascending order
+    std::sort(__for_int.begin(), __for_int.end());
+    std::sort(__time_values.begin(), __time_values.end());
+    std::sort(_for_exp.begin(), _for_exp.end());
+    integrand = elementwiseMultiply(__for_int, _for_exp);
+    // integrate the vector
+    //LinearInterpolation object with two vectors
+    LinearInterpolation integral(__time_values, integrand);
+    //integrate the vectors
+    integration_value = integral.integrate();
+  }
+
+  else 
+  {
+    integration_value = 0;
+  }
+
+  //-------------OUTPUT-------------
+  Real proportional_value = _input_signal_delayed + signalToNoise_value * 1;
+  sensor_value = drift_value + efficiency_value * (_proportional_weight * proportional_value + _integral_weight * integration_value);
+}
+
+PostprocessorValue
+GeneralSensorPostprocessor::getValue() const
+{  
+  return sensor_value;
+}
+
+void GeneralSensorPostprocessor::removeValuesBeforeTime(std::vector<Real>& _time_values, std::vector<Real>& _input_signal_values, std::vector<Real>& exp_values, Real t_desired) const 
+{
+  auto it = _time_values.begin();
+
+    // Find the iterator pointing to the element immediately before t_desired
+    auto lastBeforeTDesired = _time_values.begin();
+
+    while (it != _time_values.end())
+    {
+        if (*it < t_desired)
+        {
+            lastBeforeTDesired = it;
+            ++it;
+        }
+        else
+        {
+            // Break the loop when we find the first element greater than or equal to t_desired
+            break;
+        }
+    }

```

Comment:
could be replaced with one call of `std::upper_bound` and `std::distance`

Reply:
std::upper_bound(first, last, value) returns the iterator pointing to the first element greater than "value", but the loop i am using returns the iterator pointing to the first element lower than or equal to "value". There is another method std::lower_bound(first, last, value) but this returns the iterator pointing to the first element not less than "value". Neither upper_bound nor lower_bound does the job here.

Reply:
These lines are now 144 to 161 in the GeneralSensorPostprocessor.C file.